Some perf tweaks to the tokenizer.

emilio opened the pull request:
This makes parsing quite faster, by stripping UTF-8 logic and using tables instead of branching everywhere.

We may be able to tweak it a bit more (sometimes the table may be overkill? I don't know).

I've written the table macro so you can skip it easily if you want.

In any case, benchmark results:

Before:

> test tests::big_stylesheet ... bench:  10,392,017 ns/iter (+/- 1,954,644)
> test tests::unquoted_url   ... bench:     261,854 ns/iter (+/- 53,335)

After:

> test tests::big_stylesheet ... bench:   8,638,215 ns/iter (+/- 381,980)
> test tests::unquoted_url   ... bench:     211,863 ns/iter (+/- 73,418)

Which is quite good if you ask me.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/rust-cssparser/112)
<!-- Reviewable:end -->

emilio commented:
The stylesheet is a mix of MozReview, The Guardian, and Facebook, so I can remove it if there's any concern about having it in-repo (we can find others also).

cc @heycam @bholley, I believe this may improve our page load times, is there any way to measure Gecko's tokenizer in isolation? (to see if we should aim for more improvement or not)

r? @SimonSapin

Ms2ger commented:
> The stylesheet is a mix of MozReview, The Guardian, and Facebook, so I can remove it if there's any concern about having it in-repo (we can find others also).

Yes, please remove.

emilio commented:
> Yes, please remove.

Done

emilio commented:
Review ping @SimonSapin?

SimonSapin commented:
Reviewed 1 of 1 files at r1, 7 of 7 files at r2, 3 of 3 files at r3, 1 of 1 files at r4, 1 of 1 files at r5, 1 of 1 files at r7.
Review status: all files reviewed at latest revision, 13 unresolved discussions.

---

*[.travis.yml, line 13 at r6](https://reviewable.io:443/reviews/servo/rust-cssparser/112#-KariN9U9zn6jxTS-Sm7:-KariN9U9zn6jxTS-Sm8:bd6q46l) ([raw file](https://github.com/servo/rust-cssparser/blob/a579c022255a71e30e451aff037d161577d8f32d/.travis.yml#L13)):*
> ```YAML
>   - cargo test --features heapsize
>   - cargo test --features dummy_match_byte
>   - cargo bench --features bench
> ```

Why is it useful to run benchmarks on CI? They take time, and the CPU speed of VMs can vary a lot (e.g if another Travis job is running on the same CPU at the same time).

---

*[build.rs, line 1 at r2](https://reviewable.io:443/reviews/servo/rust-cssparser/112#-KargD9E-46brCqTK9lT:-KargD9E-46brCqTK9lU:buvxg0x) ([raw file](https://github.com/servo/rust-cssparser/blob/9c955423498898ac7a7796dbf03554cee96fd5fa/build.rs#L1)):*
> ```Rust
> 
> ```

Needs a copyright header.

---

*[src/tokenizer.rs, line 408 at r1](https://reviewable.io:443/reviews/servo/rust-cssparser/112#-KarfF5_RtRav2jbIoHi:-KarfF5_RtRav2jbIoHj:b80wyss) ([raw file](https://github.com/servo/rust-cssparser/blob/afcf259b3ae7f34e913e9b26e7ca28b06c2271c9/src/tokenizer.rs#L408)):*
> ```Rust
>         return None
>     }
>     let c = tokenizer.next_byte_unchecked();
> ```

Please rename `c` to `b` or `byte`.

---

*[src/tokenizer.rs, line 556 at r1](https://reviewable.io:443/reviews/servo/rust-cssparser/112#-Karexgv6SVBGUhJOh3V:-KarexgwPJ4hEX6Kfass:b-uabagw) ([raw file](https://github.com/servo/rust-cssparser/blob/afcf259b3ae7f34e913e9b26e7ca28b06c2271c9/src/tokenizer.rs#L556)):*
> ```Rust
>                 consume_ident_like(tokenizer)
>             } else {
>                 let ret = Delim(tokenizer.next_char());
> ```

Since we’re already excluding non-ASCII, `c as u8` can be used instead of `next_char()`. (And `ret` becomes unnecessary.)

---

*[src/tokenizer.rs, line 409 at r2](https://reviewable.io:443/reviews/servo/rust-cssparser/112#-KavEsj7_mt0oHTyJzsy:-KavEsj8ry9U9DN1sHsV:b-484s0d) ([raw file](https://github.com/servo/rust-cssparser/blob/9c955423498898ac7a7796dbf03554cee96fd5fa/src/tokenizer.rs#L409)):*
> ```Rust
>     }
>     let c = tokenizer.next_byte_unchecked();
>     let token = match_byte! { c,
> ```

Have you checked that `match_byte!` generates better code than `match`?

---

*[src/macros/match_byte.rs, line 1 at r2](https://reviewable.io:443/reviews/servo/rust-cssparser/112#-KargMI0fLUkBJj3ehGy:-KargMI1C7ZcwyoyhUly:buvxg0x) ([raw file](https://github.com/servo/rust-cssparser/blob/9c955423498898ac7a7796dbf03554cee96fd5fa/src/macros/match_byte.rs#L1)):*
> ```Rust
> 
> ```

Needs a copyright header.

---

*[src/macros/match_byte.rs, line 53 at r2](https://reviewable.io:443/reviews/servo/rust-cssparser/112#-Kav8sQTyHKkYhXf59W0:-Kav8sQUiGSG8Lj-sZrm:b3xvwat) ([raw file](https://github.com/servo/rust-cssparser/blob/9c955423498898ac7a7796dbf03554cee96fd5fa/src/macros/match_byte.rs#L53)):*
> ```Rust
> }
> 
> fn parse_match_bytes_macro(tts: Vec<syn::TokenTree>) -> (Vec<syn::TokenTree>, Vec<u8>, Vec<Case>, Option<syn::Ident>) {
> ```

This could use some docs. What is the expected "shape" of the input tokens (maybe give an example)? What are the respective components of the return value?

---

*[src/macros/match_byte.rs, line 82 at r2](https://reviewable.io:443/reviews/servo/rust-cssparser/112#-KavDS5CPQAvj470SqLQ:-KavDS5CPQAvj470SqLR:bvt8zqu) ([raw file](https://github.com/servo/rust-cssparser/blob/9c955423498898ac7a7796dbf03554cee96fd5fa/src/macros/match_byte.rs#L82)):*
> ```Rust
>     let mut tts = tts.peekable();
>     let mut case_id: u8 = 1;
>     let mut binding = None;
> ```

The parser allows multiple bindings but only the last one will work. But since this macro is not public and only used with a known set of inputs maybe that’s ok.

---

*[src/macros/match_byte.rs, line 108 at r2](https://reviewable.io:443/reviews/servo/rust-cssparser/112#-KavDA6w79aLYZqTyeus:-KavDA6xslPFVSUWboSX:b3psh4t) ([raw file](https://github.com/servo/rust-cssparser/blob/9c955423498898ac7a7796dbf03554cee96fd5fa/src/macros/match_byte.rs#L108)):*
> ```Rust
> /// The `binding` parameter is the identifier that is used by the wildcard
> /// pattern.
> fn parse_case(tts: &mut iter::Peekable<vec::IntoIter<syn::TokenTree>>,
> ```

This accepts some inputs that it shouldn’t (the empty input, an ident followed by a byte pattern, …) but since this macro is not public and only used with a known set of inputs maybe that’s ok.

---

*[src/macros/match_byte.rs, line 182 at r2](https://reviewable.io:443/reviews/servo/rust-cssparser/112#-KavDbVpCzFrOijFnRXx:-KavDbVpCzFrOijFnRXy:b5bmwcx) ([raw file](https://github.com/servo/rust-cssparser/blob/9c955423498898ac7a7796dbf03554cee96fd5fa/src/macros/match_byte.rs#L182)):*
> ```Rust
> 
>     assert!(!to_be_matched.is_empty());
>     assert_eq!(table.len(), 256);
> ```

Replace `Vec<u8>` with `[u8; 256]` to make this assert unnecessary?

---

*[src/macros/mod.rs, line 1 at r2](https://reviewable.io:443/reviews/servo/rust-cssparser/112#-KargVfncrGiNMb0CST2:-KargVfoGBZbDhU9Bc4n:buvxg0x) ([raw file](https://github.com/servo/rust-cssparser/blob/9c955423498898ac7a7796dbf03554cee96fd5fa/src/macros/mod.rs#L1)):*
> ```Rust
> 
> ```

Needs a copyright header.

---

*[src/macros/visit.rs, line 1 at r2](https://reviewable.io:443/reviews/servo/rust-cssparser/112#-Karga8ffglm6KdhooGF:-Karga8ffglm6KdhooGG:b-l1p9yj) ([raw file](https://github.com/servo/rust-cssparser/blob/9c955423498898ac7a7796dbf03554cee96fd5fa/src/macros/visit.rs#L1)):*
> ```Rust
> // Copyright 2016 The html5ever Project Developers. See the
> ```

I believe I wrote all of this code, and I agree to re-licensing it under the MPL 2. Please change this license header to the same as the rest of cssparser.

---

*[src/macros/visit.rs, line 14 at r2](https://reviewable.io:443/reviews/servo/rust-cssparser/112#-KargiUcInknqkBRrern:-KargiUcInknqkBRrero:b-1lf94a) ([raw file](https://github.com/servo/rust-cssparser/blob/9c955423498898ac7a7796dbf03554cee96fd5fa/src/macros/visit.rs#L14)):*
> ```Rust
> use syn;
> 
> pub trait Visitor {
> ```

In html5ever I initially did that but then switched to only parsing token trees and avoiding the AST entirely, so that a visitor trait is not needed. Was there a reason not to do that here?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/rust-cssparser/112)*
<!-- Sent from Reviewable.io -->

emilio commented:
I think all your comments are addressed now :)

---

Review status: 0 of 8 files reviewed at latest revision, 13 unresolved discussions.

---

*[.travis.yml, line 13 at r6](https://reviewable.io:443/reviews/servo/rust-cssparser/112#-KariN9U9zn6jxTS-Sm7:-KawLJ3A_zFt3B0zZ16g:b-g76ezj) ([raw file](https://github.com/servo/rust-cssparser/blob/a579c022255a71e30e451aff037d161577d8f32d/.travis.yml#L13)):*
<details><summary><i>Previously, SimonSapin (Simon Sapin) wrote…</i></summary><blockquote>

Why is it useful to run benchmarks on CI? They take time, and the CPU speed of VMs can vary a lot (e.g if another Travis job is running on the same CPU at the same time).
</blockquote></details>

Only to prevent them from breaking. I can back this out if you want.

---

*[build.rs, line 1 at r2](https://reviewable.io:443/reviews/servo/rust-cssparser/112#-KargD9E-46brCqTK9lT:-KawLOGylOEmh_VIPUr0:b6zrrkh) ([raw file](https://github.com/servo/rust-cssparser/blob/9c955423498898ac7a7796dbf03554cee96fd5fa/build.rs#L1)):*
<details><summary><i>Previously, SimonSapin (Simon Sapin) wrote…</i></summary><blockquote>

Needs a copyright header.
</blockquote></details>

Good catch.

---

*[src/tokenizer.rs, line 408 at r1](https://reviewable.io:443/reviews/servo/rust-cssparser/112#-KarfF5_RtRav2jbIoHi:-KawPwNKuo7zT1EqYe4o:bb6un1j) ([raw file](https://github.com/servo/rust-cssparser/blob/afcf259b3ae7f34e913e9b26e7ca28b06c2271c9/src/tokenizer.rs#L408)):*
<details><summary><i>Previously, SimonSapin (Simon Sapin) wrote…</i></summary><blockquote>

Please rename `c` to `b` or `byte`.
</blockquote></details>

Done

---

*[src/tokenizer.rs, line 556 at r1](https://reviewable.io:443/reviews/servo/rust-cssparser/112#-Karexgv6SVBGUhJOh3V:-KawLn5KYTP1PByvf2qd:b6zrrkh) ([raw file](https://github.com/servo/rust-cssparser/blob/afcf259b3ae7f34e913e9b26e7ca28b06c2271c9/src/tokenizer.rs#L556)):*
<details><summary><i>Previously, SimonSapin (Simon Sapin) wrote…</i></summary><blockquote>

Since we’re already excluding non-ASCII, `c as u8` can be used instead of `next_char()`. (And `ret` becomes unnecessary.)
</blockquote></details>

Good catch.

---

*[src/tokenizer.rs, line 409 at r2](https://reviewable.io:443/reviews/servo/rust-cssparser/112#-KavEsj7_mt0oHTyJzsy:-KawLQYaF15JsE6ZsyGf:bo6aec3) ([raw file](https://github.com/servo/rust-cssparser/blob/9c955423498898ac7a7796dbf03554cee96fd5fa/src/tokenizer.rs#L409)):*
Not manually by inspection and diffing, but I added a feature to compare both and using it made the large url test faster from:

> test tests::unquoted_url ... bench:     269,799 ns/iter (+/- 20,056)

without the custom macro expansion, to:

> test tests::unquoted_url ... bench:     213,494 ns/iter (+/- 15,135)

with it.

---

*[src/macros/match_byte.rs, line 1 at r2](https://reviewable.io:443/reviews/servo/rust-cssparser/112#-KargMI0fLUkBJj3ehGy:-KawN5s3ZE-YDobwIr1K:b-81la7d) ([raw file](https://github.com/servo/rust-cssparser/blob/9c955423498898ac7a7796dbf03554cee96fd5fa/src/macros/match_byte.rs#L1)):*
<details><summary><i>Previously, SimonSapin (Simon Sapin) wrote…</i></summary><blockquote>

Needs a copyright header.
</blockquote></details>

Right

---

*[src/macros/match_byte.rs, line 82 at r2](https://reviewable.io:443/reviews/servo/rust-cssparser/112#-KavDS5CPQAvj470SqLQ:-KawNSfHd5fs3M7jI84L:baqq4bg) ([raw file](https://github.com/servo/rust-cssparser/blob/9c955423498898ac7a7796dbf03554cee96fd5fa/src/macros/match_byte.rs#L82)):*
<details><summary><i>Previously, SimonSapin (Simon Sapin) wrote…</i></summary><blockquote>

The parser allows multiple bindings but only the last one will work. But since this macro is not public and only used with a known set of inputs maybe that’s ok.
</blockquote></details>

Yeah, I only focused in making something that kept enough idiomatic code working.

---

*[src/macros/match_byte.rs, line 182 at r2](https://reviewable.io:443/reviews/servo/rust-cssparser/112#-KavDbVpCzFrOijFnRXx:-KawQil_NsCj_bmfTPGT:bg9vyi5) ([raw file](https://github.com/servo/rust-cssparser/blob/9c955423498898ac7a7796dbf03554cee96fd5fa/src/macros/match_byte.rs#L182)):*
<details><summary><i>Previously, SimonSapin (Simon Sapin) wrote…</i></summary><blockquote>

Replace `Vec<u8>` with `[u8; 256]` to make this assert unnecessary?
</blockquote></details>

Yup, fine. I didn't want to pass it on the stack, but it's small anyway.

---

*[src/macros/visit.rs, line 1 at r2](https://reviewable.io:443/reviews/servo/rust-cssparser/112#-Karga8ffglm6KdhooGF:-KawNj2Isw7vW60qlimY:b-54avio) ([raw file](https://github.com/servo/rust-cssparser/blob/9c955423498898ac7a7796dbf03554cee96fd5fa/src/macros/visit.rs#L1)):*
<details><summary><i>Previously, SimonSapin (Simon Sapin) wrote…</i></summary><blockquote>

I believe I wrote all of this code, and I agree to re-licensing it under the MPL 2. Please change this license header to the same as the rest of cssparser.
</blockquote></details>

Ok, thanks.

---

*[src/macros/visit.rs, line 14 at r2](https://reviewable.io:443/reviews/servo/rust-cssparser/112#-KargiUcInknqkBRrern:-KawNpaNK_LbPAkph70W:b-abs9ud) ([raw file](https://github.com/servo/rust-cssparser/blob/9c955423498898ac7a7796dbf03554cee96fd5fa/src/macros/visit.rs#L14)):*
<details><summary><i>Previously, SimonSapin (Simon Sapin) wrote…</i></summary><blockquote>

In html5ever I initially did that but then switched to only parsing token trees and avoiding the AST entirely, so that a visitor trait is not needed. Was there a reason not to do that here?
</blockquote></details>

Well, we use it to visit expressions and statements in the AST. I can definitely rework it, but I think this way works quite fine.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/rust-cssparser/112#-:-KawXet5YQyr58Y4mahI:b-7lnop2)*
<!-- Sent from Reviewable.io -->

emilio commented:
Review status: 0 of 8 files reviewed at latest revision, 13 unresolved discussions.

---

*[src/tokenizer.rs, line 409 at r2](https://reviewable.io:443/reviews/servo/rust-cssparser/112#-KavEsj7_mt0oHTyJzsy:-KawYMFYDItI89x-db6n:b-z9xjaz) ([raw file](https://github.com/servo/rust-cssparser/blob/9c955423498898ac7a7796dbf03554cee96fd5fa/src/tokenizer.rs#L409)):*
<details><summary><i>Previously, emilio (Emilio Cobos Álvarez) wrote…</i></summary><blockquote>

Not manually by inspection and diffing, but I added a feature to compare both and using it made the large url test faster from:

> test tests::unquoted_url ... bench:     269,799 ns/iter (+/- 20,056)

without the custom macro expansion, to:

> test tests::unquoted_url ... bench:     213,494 ns/iter (+/- 15,135)

with it.
</blockquote></details>

After conversations in rust-internals (cc @nox), it seems that rustc doesn't generate the best possible code if we use ranges, but could if we moved the ranges to the bottom or expanded it.

We can try to make the `match_byte!` macro do any of those, and verify the performance is similar than with the table. I don't know if it's totally worth though, and I don't think I'll have the time to do it this evening :)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/rust-cssparser/112)*
<!-- Sent from Reviewable.io -->

SimonSapin commented:
Reviewed 1 of 6 files at r8, 3 of 3 files at r9, 1 of 1 files at r11, 1 of 1 files at r13, 4 of 4 files at r14, 1 of 1 files at r15.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[.travis.yml, line 13 at r6](https://reviewable.io:443/reviews/servo/rust-cssparser/112#-KariN9U9zn6jxTS-Sm7:-Kawa3RXkUfIv8LTF-P6:b6dgo1o) ([raw file](https://github.com/servo/rust-cssparser/blob/a579c022255a71e30e451aff037d161577d8f32d/.travis.yml#L13)):*
<details><summary><i>Previously, emilio (Emilio Cobos Álvarez) wrote…</i></summary><blockquote>

Only to prevent them from breaking. I can back this out if you want.
</blockquote></details>

I just checked: `cargo test` runs one iteration of benchmarks and fails if they panic. So `cargo test --features bench` does what you want but doesn’t take as much time.

---

*[src/tokenizer.rs, line 409 at r2](https://reviewable.io:443/reviews/servo/rust-cssparser/112#-KavEsj7_mt0oHTyJzsy:-KawfhlMoEkHriFdOl6x:b-e1kexd) ([raw file](https://github.com/servo/rust-cssparser/blob/9c955423498898ac7a7796dbf03554cee96fd5fa/src/tokenizer.rs#L409)):*
<details><summary><i>Previously, emilio (Emilio Cobos Álvarez) wrote…</i></summary><blockquote>

After conversations in rust-internals (cc @nox), it seems that rustc doesn't generate the best possible code if we use ranges, but could if we moved the ranges to the bottom or expanded it.

We can try to make the `match_byte!` macro do any of those, and verify the performance is similar than with the table. I don't know if it's totally worth though, and I don't think I'll have the time to do it this evening :)
</blockquote></details>

This is already an improvement that I’m happy with, so up to you to land it as-is or investigate alternatives first.

---

*[src/macros/match_byte.rs, line 182 at r2](https://reviewable.io:443/reviews/servo/rust-cssparser/112#-KavDbVpCzFrOijFnRXx:-Kawfxfqm1TBlvrwKWlk:bqzijt2) ([raw file](https://github.com/servo/rust-cssparser/blob/9c955423498898ac7a7796dbf03554cee96fd5fa/src/macros/match_byte.rs#L182)):*
<details><summary><i>Previously, emilio (Emilio Cobos Álvarez) wrote…</i></summary><blockquote>

Yup, fine. I didn't want to pass it on the stack, but it's small anyway.
</blockquote></details>

Could also be `Box<[u8; 256]>`. Anyway, this is compile-time code that is only used a few times and with small inputs, so I’m not worried about its perfomance.

---

*[src/macros/visit.rs, line 14 at r2](https://reviewable.io:443/reviews/servo/rust-cssparser/112#-KargiUcInknqkBRrern:-KawgWTfFikqqM1xE4tz:b-uxqw50) ([raw file](https://github.com/servo/rust-cssparser/blob/9c955423498898ac7a7796dbf03554cee96fd5fa/src/macros/visit.rs#L14)):*
<details><summary><i>Previously, emilio (Emilio Cobos Álvarez) wrote…</i></summary><blockquote>

Well, we use it to visit expressions and statements in the AST. I can definitely rework it, but I think this way works quite fine.
</blockquote></details>

Yeah, both work. It’s just less code to maintain, or to update if the AST needs to change: https://github.com/servo/html5ever/commit/2ddcd53fac6fa2310cc39bb8a3afe2b155a2e787

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/rust-cssparser/112)*
<!-- Sent from Reviewable.io -->

emilio commented:
Review status: 4 of 7 files reviewed at latest revision, 2 unresolved discussions.

---

*[.travis.yml, line 13 at r6](https://reviewable.io:443/reviews/servo/rust-cssparser/112#-KariN9U9zn6jxTS-Sm7:-Kb07f07CelaC_FfmV-2:b-px11i4) ([raw file](https://github.com/servo/rust-cssparser/blob/a579c022255a71e30e451aff037d161577d8f32d/.travis.yml#L13)):*
<details><summary><i>Previously, SimonSapin (Simon Sapin) wrote…</i></summary><blockquote>

I just checked: `cargo test` runs one iteration of benchmarks and fails if they panic. So `cargo test --features bench` does what you want but doesn’t take as much time.
</blockquote></details>

Oh, right! done :)

---

*[src/tokenizer.rs, line 409 at r2](https://reviewable.io:443/reviews/servo/rust-cssparser/112#-KavEsj7_mt0oHTyJzsy:-Kb07HQG2EVLgC1K5l6z:b5ec7nh) ([raw file](https://github.com/servo/rust-cssparser/blob/9c955423498898ac7a7796dbf03554cee96fd5fa/src/tokenizer.rs#L409)):*
<details><summary><i>Previously, SimonSapin (Simon Sapin) wrote…</i></summary><blockquote>

This is already an improvement that I’m happy with, so up to you to land it as-is or investigate alternatives first.
</blockquote></details>

Ok, I think I want to land this as-is for now.

---

*[src/macros/match_byte.rs, line 182 at r2](https://reviewable.io:443/reviews/servo/rust-cssparser/112#-KavDbVpCzFrOijFnRXx:-Kb07R3ZIlI_qwum2CXA:bgc2iqb) ([raw file](https://github.com/servo/rust-cssparser/blob/9c955423498898ac7a7796dbf03554cee96fd5fa/src/macros/match_byte.rs#L182)):*
<details><summary><i>Previously, SimonSapin (Simon Sapin) wrote…</i></summary><blockquote>

Could also be `Box<[u8; 256]>`. Anyway, this is compile-time code that is only used a few times and with small inputs, so I’m not worried about its perfomance.
</blockquote></details>

Yup, agreed.

---

*[src/macros/visit.rs, line 14 at r2](https://reviewable.io:443/reviews/servo/rust-cssparser/112#-KargiUcInknqkBRrern:-Kb07T3YCgCyY_3stHak:bbk9wde) ([raw file](https://github.com/servo/rust-cssparser/blob/9c955423498898ac7a7796dbf03554cee96fd5fa/src/macros/visit.rs#L14)):*
<details><summary><i>Previously, SimonSapin (Simon Sapin) wrote…</i></summary><blockquote>

Yeah, both work. It’s just less code to maintain, or to update if the AST needs to change: https://github.com/servo/html5ever/commit/2ddcd53fac6fa2310cc39bb8a3afe2b155a2e787
</blockquote></details>

Thanks for the pointers :)

I made it work on token trees, I've left the last fixup commit intentionally so you can review it.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/rust-cssparser/112)*
<!-- Sent from Reviewable.io -->

SimonSapin commented:
Squash as desired, then r=me :)

---

Reviewed 1 of 1 files at r16, 3 of 3 files at r17.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/servo/rust-cssparser/112#-:-Kb1cgfxov6D81txXBm0:by4t76t)*
<!-- Sent from Reviewable.io -->

emilio commented:
@bors-servo r=SimonSapin

bors-servo commented:
:pushpin: Commit 21f8573 has been approved by `SimonSapin`

<!-- @bors-servo r=SimonSapin 21f85731aa14ccc0a1f75da9ae31c66b2a88654f -->

bors-servo commented:
:hourglass: Testing commit 21f8573 with merge 7c8a9a6...

bors-servo commented:
:sunny: Test successful - [status-travis](https://travis-ci.org/servo/rust-cssparser/builds/194076634)

emilio commented:
Thanks for the review @SimonSapin! :)

nox commented:
Part of this (or maybe all of it) can be reverted once https://github.com/rust-lang/rust/pull/39456 is in Rust stable.

SimonSapin commented:
#128
